### DISCUSSION TITLE
Why memcpy kernels are submitted to command queue one by one?

### DISCUSSION THREAD

**Discussion by GHGmc2** [intel/llvm#4034] [Q&A]
With [ze_tracer](https://github.com/intel/pti-gpu/tree/master/tools/ze_tracer) tool, I noticed that memcpy kernels in the command list are submited to command queue one by one(I set `SYCL_PI_LEVEL_ZERO_BATCH_SIZE=32` and did not call wait() after the memcpy), and there are bubbles between the pipeline.

**Why can't we submit them in mini-batch, and is there any change to minimize the bubble between them?**

- code of host2device memcpy, the same to d2d and d2h memcpy.
```c++
#include "CL/sycl.hpp"

constexpr int ALIGN = 64;
constexpr int SIZE = 256;

int main() {
  sycl::queue q(sycl::gpu_selector{});

  for (int i = 0; i < 20; ++i) {
    void *src = sycl::aligned_alloc_host(ALIGN, SIZE, q);
    void *dst = sycl::aligned_alloc_device(ALIGN, SIZE, q);
    q.memcpy(dst, src, SIZE);
    sycl::free(dst, q);
    sycl::free(src, q);
  }

  return 0;
}
```

- the timeline the ze_tracer generates(w/ flag `--chrome-device-stages`)
![image](https://user-images.githubusercontent.com/6487326/124121048-03053a80-daa7-11eb-8b64-270dd5932553.png)

**Comment by AlexeySachkov** [comment]
Tagging @smaslov-intel and @kbsmith-intel here to help answering this question

**Reply by smaslov-intel** [reply]
That's right, we are not currently batching memory transfers, and this is WIP to start doing so.
This would save on the submissions time, since there would be a single submission then, and is likely to get rid of bubbles too.

**Reply by GHGmc2** [reply]
Thanks, will memcpy be submitted with other normal kernels in one batch, or only batch of memcpy can be submitted together?
